cli_tools/google-osconfig-agent/ospatch/patch_run.go

```diff
@@ -356,50 +340,66 @@ func ackPatch(ctx context.Context, patchJobName string) {
 			return
 		}
-		tasker.Enqueue("Run patch", func() {
-			patchRunner(ctx, &patchRun{Job: &patchJob{res}})
-		})
+		pr.PatchStep = started
+		tasker.Enqueue("Run patch", pr.runPatch)
 	}
 }
 
-func reportPatchDetails(ctx context.Context, patchJobName string, patchState osconfigpb.Instance_PatchState, attemptCount int64, failureReason string) (*osconfigpb.ReportPatchJobInstanceDetailsResponse, error) {
-	// TODO: add retries. Patching shouldn't continue if we can't talk to the server.
-	logger.Debugf("Reporting patch details name:'%s', state:'%s', failReason:'%s'", patchJobName, patchState, failureReason)
-
-	client, err := osconfig.NewClient(ctx, option.WithEndpoint(config.SvcEndpoint()), option.WithCredentialsFile(config.OAuthPath()))
-	if err != nil {
-		logger.Errorf("osconfig.NewClient Error: %v", err)
-		return nil, err
-	}
-	defer client.Close()
-
-	// This can't be cached.
-	identityToken, err := metadata.Get(identityTokenPath)
-	if err != nil {
-		return nil, err
-	}
+// reportPatchDetails tries to report patch details for no more than 2109s (really only about ~30min).
+func (r *patchRun) reportPatchDetails(patchState osconfigpb.Instance_PatchState, attemptCount int64, failureReason string) error {
+	var err error
+	rnd := rand.New(rand.NewSource(time.Now().UnixNano()))
+	var tot int
+	for i := 1; tot < 1800; i++ {
+		err := func() error {
+			// This can't be cached.
+			identityToken, err := metadata.Get(identityTokenPath)
+			if err != nil {
+				return err
+			}
 
-	request := osconfigpb.ReportPatchJobInstanceDetailsRequest{
-		Resource:         config.Instance(),
-		InstanceSystemId: config.ID(),
-		PatchJobName:     patchJobName,
-		InstanceIdToken:  identityToken,
-		State:            patchState,
-		AttemptCount:     attemptCount,
-		FailureReason:    failureReason,
-	}
+			request := osconfigpb.ReportPatchJobInstanceDetailsRequest{
+				Resource:         config.Instance(),
+				InstanceSystemId: config.ID(),
+				PatchJobName:     r.Job.PatchJobName,
+				InstanceIdToken:  identityToken,
+				State:            patchState,
+				AttemptCount:     attemptCount,
+				FailureReason:    failureReason,
+			}
+			logger.Debugf("Reporting patch details request: %+v", request)
 
-	res, err := client.ReportPatchJobInstanceDetails(ctx, &request)
-	if err != nil {
-		if s, ok := status.FromError(err); ok {
-			return nil, fmt.Errorf("code: %q, message: %q, details: %q", s.Code(), s.Message(), s.Details())
+			res, err := r.client.ReportPatchJobInstanceDetails(r.ctx, &request)
+			if err != nil {
+				if s, ok := status.FromError(err); ok {
+					return fmt.Errorf("code: %q, message: %q, details: %q", s.Code(), s.Message(), s.Details())
+				}
+				return err
+			}
+			r.Job.ReportPatchJobInstanceDetailsResponse = res
+			return nil
+		}()
+		if err == nil {
+			return nil
 		}
-		return nil, err
+		logger.Debugf("Error repoting patch details, attempt %d: %v", i, err)
+
+		// Always increasing with some jitter, longest wait will be ~5min.
```

Comment:
Consider clamping this wait time to be sure in case `tot` ever changes, eg. `sleepTime = min(i * i, 300) + jitter`